numpy/_core/arrayprint.py

```diff
@@ -394,12 +380,12 @@ def printoptions(*args, **kwargs):
     set_printoptions, get_printoptions
 
     """
-    opts = np.get_printoptions()
     try:
-        np.set_printoptions(*args, **kwargs)
-        yield np.get_printoptions()
+        token = set_printoptions(*args, **kwargs)
+        yield get_printoptions()
     finally:
-        np.set_printoptions(**opts)
+        _format_options.reset(token)
+        _set_legacy_print_mode(_format_options.get())
```

Comment:
Is this necessary? Shouldn't the old state be guaranteed-correct?

Reply:
The original `np.set_printoptions(**opts)` here did C variable update as a side effect. `_format_options.reset(token)` only updates the context variable so I think we need another function call to also reflect the `legacy_print_mode` update.

Reply:
But the C variable is no longer used in this PR. And `_set_legacy_print_mode` only possibly sets the sign option. But that can never have been out of whack in stored previous options.

Reply:
I think this is correct now.